### PR TITLE
DAS Sampler

### PR DESCRIPTION
**Pull request by zilm13**
Mostly copying DasCustody logic in implementation but made separately to add more flexibility when needed.
- refactored `DataColumnSidecarCustodyImpl`, some common inner classes moved out
- `firstIncompleteSlot` -> `firstCustodyIncompleteSlot`, for sampler separate variable - `firstSamplerIncompleteSlot`
- non-DAS broken tests fixed/disabled
- logging when sampling of slot is completed. In future it could be callback for forkchoice or whatever
- most new logic is in `DataColumnSidecarSamplerImpl`

### CODE REVIEW

**Comment by zilm13** [comment]
Updated with fixes

File: ethereum/statetransition/src/main/java/tech/pegasys/teku/statetransition/datacolumns/DasSamplerSyncImpl.java [hunk 1]
```diff
@@ -0,0 +1,148 @@
...
+import java.util.stream.Collectors;
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import tech.pegasys.teku.ethereum.events.SlotEventsChannel;
+import tech.pegasys.teku.infrastructure.async.SafeFuture;
+import tech.pegasys.teku.infrastructure.unsigned.UInt64;
+import tech.pegasys.teku.spec.datastructures.blobs.versions.eip7594.DataColumnSidecar;
+import tech.pegasys.teku.statetransition.datacolumns.retriever.DataColumnSidecarRetriever;
+
+public class DasSamplerSyncImpl implements SlotEventsChannel, DasSamplerSync {
```
Comment (Nashatyrev) [inline ethereum/statetransition/src/main/java/tech/pegasys/teku/statetransition/datacolumns/DasSamplerSyncImpl.java]: nit: the class is almost identical to `DasCustodySync`. Probably subject to refactor in production

**Review by Nashatyrev — APPROVED** [review]
LGTM

**Comment by Nashatyrev** [comment]
@zilm13 sorry forgot a couple of proposed changes: 
```patch
Index: ethereum/statetransition/src/main/java/tech/pegasys/teku/statetransition/datacolumns/DasSamplerImpl.java
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/ethereum/statetransition/src/main/java/tech/pegasys/teku/statetransition/datacolumns/DasSamplerImpl.java b/ethereum/statetransition/src/main/java/tech/pegasys/teku/statetransition/datacolumns/DasSamplerImpl.java
--- a/ethereum/statetransition/src/main/java/tech/pegasys/teku/statetransition/datacolumns/DasSamplerImpl.java	(revision a4e4fe1c8f0195e2d295abdb681364118045960f)
+++ b/ethereum/statetransition/src/main/java/tech/pegasys/teku/statetransition/datacolumns/DasSamplerImpl.java	(date 1721226597191)
@@ -474,7 +474,7 @@
     if (assignedSampleColumns.isEmpty()) {
       return Collections.emptyList();
     }
-    SlotAndBlockRoot from = assignedSampleColumns.navigableKeySet().getFirst();
+    SlotAndBlockRoot from = assignedSampleColumns.navigableKeySet().first();
     final UInt64 toSlot;
     if (from.getSlot().plus(limit).isLessThan(toSlotIncluded)) {
       toSlot = from.getSlot().plus(limit);
@@ -493,7 +493,7 @@
         .toList();
   }
 
-  public SafeFuture<Void> addSlotTask(
+  private SafeFuture<Void> addSlotTask(
       final UInt64 slot, final Bytes32 blockRoot, final Bytes32 parentRoot) {
     return assignSampleColumns(slot, blockRoot, parentRoot);
   }
```

**Comment by Nashatyrev** [comment]
Pushed directly to `das` as it didn't compile: 1a0e3535d13b68484f44d6e19c27fa01b81d9451

**Pull request merged**